File: scripts/submit-indexing.mjs
#!/usr/bin/env node
/**
 * Google Search Console URL Indexing Script
 * ==========================================
 * Submits URLs to Google for indexing via the Indexing API.
 *
 * SETUP:
 * 1. Go to https://console.cloud.google.com/
 * 2. Create a project (or use existing)
 * 3. Enable "Web Search Indexing API" (aka Indexing API)
 * 4. Go to IAM & Admin > Service Accounts > Create Service Account
 * 5. Create a JSON key and save it as: scripts/gsc-service-account.json
 * 6. Copy the service account email (e.g., name@project.iam.gserviceaccount.com)
 * 7. In Google Search Console > Settings > Users and permissions > Add user
 *    Add the service account email as "Owner"
 * 8. Run: node scripts/submit-indexing.mjs [--day N] [--dry-run]
 *
 * USAGE:
 *   node scripts/submit-indexing.mjs              # Submit today's batch
 *   node scripts/submit-indexing.mjs --day 1      # Submit day 1 batch
 *   node scripts/submit-indexing.mjs --dry-run    # Preview without submitting
 *   node scripts/submit-indexing.mjs --status      # Show progress summary
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const QUEUE_PATH = join(__dirname, 'indexing-queue.json');
const PROGRESS_PATH = join(__dirname, 'indexing-progress.json');
const CREDENTIALS_PATH = join(__dirname, 'gsc-service-account.json');

// ─── Parse CLI args ──────────────────────────────────────────────────────────

const args = process.argv.slice(2);
const dryRun = args.includes('--dry-run');
const showStatus = args.includes('--status');
const dayIndex = args.indexOf('--day');
const specifiedDay = dayIndex !== -1 ? parseInt(args[dayIndex + 1], 10) : null;

// ─── Load data ───────────────────────────────────────────────────────────────

function loadQueue() {
  if (!existsSync(QUEUE_PATH)) {
    console.error('ERROR: indexing-queue.json not found. Run generate-indexing-queue.py first.');
    process.exit(1);
  }
  return JSON.parse(readFileSync(QUEUE_PATH, 'utf-8'));
}

function loadProgress() {
  if (!existsSync(PROGRESS_PATH)) {
    return { submitted: {}, lastDay: 0 };
  }
  return JSON.parse(readFileSync(PROGRESS_PATH, 'utf-8'));
}

function saveProgress(progress) {
  writeFileSync(PROGRESS_PATH, JSON.stringify(progress, null, 2), 'utf-8');
}

// ─── Google Indexing API ─────────────────────────────────────────────────────

async function getAccessToken(credentials) {
  // Create JWT
  const header = { alg: 'RS256', typ: 'JWT' };
  const now = Math.floor(Date.now() / 1000);
  const claim = {
    iss: credentials.client_email,
    scope: 'https://www.googleapis.com/auth/indexing',
    aud: 'https://oauth2.googleapis.com/token',
    exp: now + 3600,
    iat: now,
  };

  const { createSign } = await import('crypto');

  const encode = (obj) =>
    Buffer.from(JSON.stringify(obj)).toString('base64url');

  const unsignedToken = `${encode(header)}.${encode(claim)}`;

  const sign = createSign('RSA-SHA256');
  sign.update(unsignedToken);
  const signature = sign.sign(credentials.private_key, 'base64url');

  const jwt = `${unsignedToken}.${signature}`;

  // Exchange JWT for access token
  const res = await fetch('https://oauth2.googleapis.com/token', {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: `grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer&assertion=${jwt}`,
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Token exchange failed: ${res.status} ${text}`);
  }

  const data = await res.json();
  return data.access_token;
}

async function submitUrl(url, accessToken) {
  const res = await fetch(
    'https://indexing.googleapis.com/v3/urlNotifications:publish',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${accessToken}`,
      },
      body: JSON.stringify({ url, type: 'URL_UPDATED' }),
    }
  );

  const data = await res.json();
  return { status: res.status, data };
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ─── Main ────────────────────────────────────────────────────────────────────

async function main() {
  const queue = loadQueue();
  const progress = loadProgress();

  // Status mode
  if (showStatus) {
    const submitted = Object.keys(progress.submitted).length;
    const total = queue.new_pages_count;
    console.log(`\nIndexing Progress: ${submitted}/${total} URLs submitted`);
    console.log(`Last batch day: ${progress.lastDay}`);
    console.log(`Remaining: ${total - submitted} URLs over ${queue.total_days_needed - progress.lastDay} days\n`);

    for (const batch of queue.batches) {
      const done = batch.urls.filter((u) => progress.submitted[u]).length;
      const icon = done === batch.urls.length ? 'DONE' : done > 0 ? 'PARTIAL' : 'PENDING';
      console.log(
        `  Day ${batch.day} (${batch.date}) [${icon}]: ${done}/${batch.urls.length} - ${batch.priority}`
      );
    }
    return;
  }

  // Determine which day to submit
  let targetDay;
  if (specifiedDay) {
    targetDay = specifiedDay;
  } else {
    // Find today's date in the queue
    const today = new Date().toISOString().split('T')[0];
    const todayBatch = queue.batches.find((b) => b.date === today);
    if (todayBatch) {
      targetDay = todayBatch.day;
    } else {
      // Use next unsubmitted day
      targetDay = progress.lastDay + 1;
    }
  }

  const batch = queue.batches.find((b) => b.day === targetDay);
  if (!batch) {
    console.log(`No batch found for day ${targetDay}. All ${queue.total_days_needed} days may be complete.`);
    return;
  }

  // Filter out already-submitted URLs
  const toSubmit = batch.urls.filter((u) => !progress.submitted[u]);

  console.log(`\n=== Day ${batch.day} (${batch.date}) - ${batch.priority} ===`);
  console.log(`URLs in batch: ${batch.urls.length}`);
  console.log(`Already submitted: ${batch.urls.length - toSubmit.length}`);
  console.log(`To submit: ${toSubmit.length}`);
  console.log(`Mode: ${dryRun ? 'DRY RUN' : 'LIVE'}\n`);

  if (toSubmit.length === 0) {
    console.log('All URLs in this batch already submitted. Use --day N for another batch.');
    return;
  }

  if (dryRun) {
    for (const url of toSubmit) {
      console.log(`  [DRY RUN] Would submit: ${url}`);
    }
    console.log(`\nDry run complete. ${toSubmit.length} URLs would be submitted.`);
    return;
  }

  // Load credentials
  if (!existsSync(CREDENTIALS_PATH)) {
    console.error('ERROR: gsc-service-account.json not found.');
    console.error('See setup instructions at the top of this file.');
    console.error('\nRunning in dry-run mode instead:\n');
    for (const url of toSubmit) {
      console.log(`  [NO CREDS] Would submit: ${url}`);
    }
    return;
  }

  const credentials = JSON.parse(readFileSync(CREDENTIALS_PATH, 'utf-8'));
  console.log(`Authenticating as: ${credentials.client_email}\n`);

  let accessToken;
  try {
    accessToken = await getAccessToken(credentials);
  } catch (err) {
    console.error(`Authentication failed: ${err.message}`);
    process.exit(1);
  }

  let success = 0;
  let failed = 0;

  for (const url of toSubmit) {
    try {
      const result = await submitUrl(url, accessToken);
      if (result.status === 200) {
        console.log(`  OK: ${url}`);
        progress.submitted[url] = new Date().toISOString();
        success++;
      } else {
        console.log(`  FAIL (${result.status}): ${url} - ${JSON.stringify(result.data)}`);
        failed++;
      }
    } catch (err) {
      console.log(`  ERROR: ${url} - ${err.message}`);
      failed++;
    }

    // Rate limit: 1 second between requests
    await sleep(1000);
  }

  progress.lastDay = Math.max(progress.lastDay, targetDay);
  saveProgress(progress);

  console.log(`\n=== Summary ===`);
  console.log(`Submitted: ${success}`);
  console.log(`Failed: ${failed}`);
  console.log(`Total progress: ${Object.keys(progress.submitted).length}/${queue.new_pages_count}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/submit-indexing.py
#!/usr/bin/env python3
"""
Submit URLs to the Google Indexing API (URL_UPDATED) across all available
service accounts.  RUN AFTER DEPLOY ONLY — submitting URLs that are not yet
live makes Google crawl 404s / stale content.

Caps: 200 URLs per service account per day, global cap = 200 * (#accounts).
With 10 accounts that is 2000 URLs/day.

Usage:
    python3 submit-indexing.py seo-changed-urls.txt            # live submit
    python3 submit-indexing.py seo-changed-urls.txt --dry-run  # preview only

Requires: pip install google-api-python-client google-auth
The service-account JSON files (gsc-service-account-*.json) must live in this
scripts/ folder and be added as Owners of the GSC property, with the
"Indexing API" enabled in their Google Cloud projects.
"""
import sys, os, json, time, glob, datetime

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
SA_GLOB = os.path.join(SCRIPT_DIR, "gsc-service-account-*.json")
PER_ACCOUNT_DAILY = 200
SCOPES = ["https://www.googleapis.com/auth/indexing"]
LEDGER = os.path.join(SCRIPT_DIR, ".indexing-submit-ledger.jsonl")
STATE  = os.path.join(SCRIPT_DIR, ".indexing-daily-state.json")


def load_state():
    today = str(datetime.date.today())
    if os.path.exists(STATE):
        try:
            s = json.load(open(STATE))
            if s.get("day") == today:
                return s
        except Exception:
            pass
    return {"day": today, "perAccount": {}}


def main():
    from google.oauth2 import service_account
    from googleapiclient.discovery import build

    argv = sys.argv[1:]
    dry = "--dry-run" in argv
    files = [a for a in argv if not a.startswith("--")]
    if not files:
        print(__doc__); return
    urls = [u.strip() for u in open(files[0], encoding="utf-8") if u.strip() and not u.startswith("#")]
    # de-dupe, keep order
    seen = set(); urls = [u for u in urls if not (u in seen or seen.add(u))]

    sas = sorted(glob.glob(SA_GLOB))
    if not sas:
        print("No service-account files found at", SA_GLOB); return

    state = load_state()
    cap = PER_ACCOUNT_DAILY * len(sas)
    print(f"{len(urls)} URLs | {len(sas)} accounts | daily cap {cap} | already used today: {sum(state['perAccount'].values())}")

    svcs = []
    for f in sas:
        try:
            creds = service_account.Credentials.from_service_account_file(f, scopes=SCOPES)
            svcs.append((os.path.basename(f), build("indexing", "v3", credentials=creds, cache_discovery=False)))
        except Exception as e:
            print("  skip", os.path.basename(f), str(e)[:100])
    if not svcs:
        print("No usable service accounts."); return

    led = open(LEDGER, "a", encoding="utf-8")
    sent = 0; i = 0
    for url in urls:
        if sent >= cap:
            print("Global daily cap reached."); break
        placed = False
        for _ in range(len(svcs)):
            name, svc = svcs[i % len(svcs)]; i += 1
            if state["perAccount"].get(name, 0) >= PER_ACCOUNT_DAILY:
                continue
            if dry:
                print(f"[dry] {url}  ->  {name}")
            else:
                try:
                    svc.urlNotifications().publish(body={"url": url, "type": "URL_UPDATED"}).execute()
                    led.write(json.dumps({"ts": datetime.datetime.utcnow().isoformat(), "url": url, "sa": name}) + "\n")
                except Exception as e:
                    print("  ERR", url, name, str(e)[:120]); continue
            state["perAccount"][name] = state["perAccount"].get(name, 0) + 1
            sent += 1; placed = True; break
        if not placed:
            print(f"All accounts exhausted today after {sent} URLs."); break
        if not dry:
            time.sleep(0.3)
    json.dump(state, open(STATE, "w"), indent=1); led.close()
    print(f"Done. {'(dry-run) ' if dry else ''}Submitted {sent} URLs.")


if __name__ == "__main__":
    main()
